### PR TITLE
i#6471 sched idle: Add idle time from blocking syscalls

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a longer waiting period on blocking syscalls using either provided output time or for instruction-based quanta a count of queue selections before a blocked input is actually selected.  Since the scheduler does not have timer interrupts or regular points of control and relies on its user calling it, idle inputs are kept on the ready queue and are checked for becoming unblocked when the ready queue is queried.

The wait duration is set based on the "wait time factor" which is the syscall latency divided by the context switch threshold multipled by a user-provided "block_time_scale" option which can be used to scale up or down the durations.

The wait duration is erased on a direct switch to an input.

Adds a new replay record type to represent idle time on replay.

Augments the unit tests to include blocking high-latency syscalls to test the new feature in various sub-tests.

Issue: #6471

### CODE REVIEW

File: clients/drcachesim/scheduler/scheduler.h [hunk 1]
```diff
@@ -522,6 +522,17 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
          * blocking and trigger a context switch.
          */
         uint64_t blocking_switch_threshold = 100;
+        /**
+         * Controls the amount of time inputs are considered blocked at a syscall whose
+         * latency exceeds #syscall_switch_threshold or #blocking_switch_threshold.  A
+         * "block time factor" is computed from the syscall latency divided by either
+         * #syscall_switch_threshold or #blocking_switch_threshold.  This factor is
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.h:529]: These thresholds are applied in raw2trace whereas the block_time_scale is applied post-raw2trace. If the thresholds change after raw2trace (and therefore the maybe_blocking markers) would that be okay?

File: clients/drcachesim/scheduler/scheduler.h [hunk 2]
```diff
@@ -522,6 +522,17 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
...
+         * multiplied by this field #block_time_scale.  For #QUANTUM_TIME, that amount of
+         * time, as reported by the time parameter to next_record(), must pass before the
+         * input is no longer considered blocked.  For instruction quanta, that count of
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.h:532]: Do you instead mean scheduler selections equal to the instruction_quanta times the factor must occur?

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 3]
```diff
@@ -928,7 +940,7 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_traced_schedule()
         // a signal can come in after a prior timestamp with no intervening instrs).
         if (!all_sched[cur_output].empty() &&
             input == all_sched[cur_output].back().key.input &&
-            start == all_sched[cur_output].back().start_instruction) {
+            start == all_sched[cur_output].back().value.start_instruction) {
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:943]: Do we need to confirm whether all_sched[cur_output].back() is not an IDLE entry? Same elsewhere.

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 4]
```diff
@@ -744,7 +744,12 @@ scheduler_tmpl_t<RecordType, ReaderType>::set_initial_schedule(
             }
             for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
                 if (i < static_cast<input_ordinal_t>(inputs_.size())) {
-                    input_info_t *queue_next = pop_from_ready_queue(i);
+                    input_info_t *queue_next;
+#ifdef DEBUG
+                    sched_type_t::stream_status_t status =
+#endif
+                        pop_from_ready_queue(i, queue_next);
+                    assert(status == STATUS_OK); // No blocked inputs yet.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/scheduler/scheduler.cpp:752]: But wait... asserts still work in non-debug builds, so how is this building successfully if there's no `status` in non-debug.

File: clients/drcachesim/scheduler/scheduler.h (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:532]: Yes, "that count" is the final number referenced at line 530 "This factor is multiplied by this field block_time_scale".  Reworded to call it "a final value" and s/that count/that final value/ so it's clearer.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:943]: Here it cannot be, by construction: this is an as-traced schedule from which we construct all_sched from only DEFAULT types.  I added an assert here as a sanity check and another one in the check_and_fix_* routine.

**Comment by derekbruening** [comment]
In further testing on large traces, I am actually having trouble getting enough idle time.  I can get a number of idle periods but they never add up to a large % of the instruction count.  (With large core count I can get plenty of idle time.).  I can't get arbitrary idle times by raising the scale: the scheduler gets slow before I reach idle % targets.  But, I think it is worth merging this as-is for now to get some base blocked-idle time in there for base testing, even if I figure out how to improve it later.

**Comment by derekbruening** [comment]
x32 is amd #6417 
win64 is windows-timestamps callback.c assert #6081

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:752]: Grr, on Github there are no builds that have assert enabled but don't have DEBUG.  But in our internal builds, we do have such a mode (and my pre-merge testing, which was done here, only tests release build and so didn't see it).

**Pull request merged**